### ISSUE TITLE
How to host a WPF control 

### ISSUE THREAD

**Issue by andreinitescu** [mono/xwt#515]
I'm completely new to XWT, is it possible to host a WPF Control in a XWT widget?
Obviously, this is on Windows.

**Comment by umbreon222** [comment]
I'm almost positive that you can host a wpf control within xwt. The method in which you do so would most likely rely on the context of the situation. What wpf control were you planning on hosting? Also as a side note, the only benefit of xwt is cross platform compatability. Using the wpf and limiting yourself to Windows is kind of rhetorical don't you think?

**Comment by andreinitescu** [comment]
wow, thanks for your message.
I'm trying to implement a XML editing feature, so I need a XWT control which is a XML editor.
So that's why I thought with WPF hosting would be easier. 
Is there any XWT XML editor control?

**Comment by umbreon222** [comment]
As far as I can tell from a quick glance, [this](https://github.com/mono/xwt/tree/master/Xwt/Xwt) is a link to the current controls of XWT. It seems there is no XML editing controls as of now, even after a quick Google search. If you want the control to be stable I would just switch the whole project to WPF if it is easy to convert and use [this](https://wpfxmleditor.codeplex.com/) example which is an XML editor in WPF. If using the cross platform capability of XWT is a must, I am sure it could be converted with some time and effort. Although, converting it to XWT doesn't guarantee stability. I hope that helps haha.

**Comment by andreinitescu** [comment]
I'm new to XWT but it doesn't like that there is any XWT community...

**Comment by umbreon222** [comment]
Community is a bit sparse with the XWT project and documentations is even more sparse. lol

**Comment by sevoku** [comment]
take look at https://github.com/luiscubal/XwtPlus.TextEditor, may be a good starting point.

**Comment by cra0zy** [comment]
XwtPlus.TextEditor bug list:
- broken input on non WPF backends (#384 should fix up Gtk)
- broken selection
- line numbers after selected line get wrong color
- no cut/copy/paste
- changing to a different theme results in a really broken theme

I've fixed everything besides selection and themes locally and this is how it ended up looking: https://youtu.be/ThdzkckrJkU xml tab.

If you really want to use a WPF control inside Xwt the good starting point is:
```c#
Xwt.WPFBackend.WPFEngine we = new Xwt.WPFBackend.WPFEngine();
we.GetBackendForContext ???
```
I've never converted native widget to xwt widget, so I don't know how it really works, but I've done the opposite(converted xwt widget to native widget).

Side note: @sevoku will the XwtPlus.TextEditor ever be moved into the main xwt repo?

**Comment by sevoku** [comment]
XwtPlus.TextEditor is far from being usable. Maybe you can contact @luiscubal for more information. If you have managed to fix some bugs, it would be cool if you could fork and push them to github.

About hosting a native control: I don't know how it can be done easily. It should be possible (maybe there  are some examples in MonoDevelop sources) but the "right" way is like its done for all other widgets in core Xwt or the other way round like XwtPlus.TextEditor.
If you want Xwt to use native Xml/Text editor controls, then you have to create the frontend widget and backends for every platform (some Xwt Wpf control for Wpf, maybe MonoDevelop.TextEditor for Gtk,...). The other way is not to use any native controls and let Xwt do the job (just like XwtPlus.TextEditor).

I don't think that it would ever become part of the official Xwt. It really complex and needs much maintenance, so a separate project is a good place for it.

**Comment by andreinitescu** [comment]
thanks all, I figured it out I can use something else.
thanks again

**Comment by cra0zy** [comment]
Just FYI, I've finally had time to organize all the the patches for XwtPlus.TextEditor. You can get it from: https://github.com/cra0zy/XwtPlus.TextEditor, everything besides theming is fixed.

I've also submitted a PR to the original repo: https://github.com/luiscubal/XwtPlus.TextEditor/pull/1

**Comment by sevoku** [comment]
I've merged #384, so this should work with Gtk now.
